fetch/src/main/java/com/indix/gocd/s3fetch/FetchExecutor.java

```diff
@@ -14,37 +16,82 @@
-        final FetchConfig fetchConfig = getFetchConfig(config, context);
 
-        ValidationResult validationResult = fetchConfig.validate();
-        if(!validationResult.isSuccessful()) {
-            return ExecutionResult.failure(validationResult.getMessages().toString());
-        }
+        try {
+            final FetchConfig fetchConfig = getFetchConfig(config, context);
 
-        final S3ArtifactStore store = s3ArtifactStore(fetchConfig);
+            ValidationResult validationResult = fetchConfig.validate();
+            if (!validationResult.isSuccessful()) {
+                logger.error(String.format("s3 fetch configuration is invalid: %s", validationResult.getMessages().toString()));
+                return ExecutionResult.failure(validationResult.getMessages().toString());
+            }
+            final S3ArtifactStore store = s3ArtifactStore(fetchConfig);
 
-        String artifactPathOnS3 = fetchConfig.getArtifactsLocationTemplate();
-        context.console().printLine(String.format("Getting artifacts from %s", store.pathString(artifactPathOnS3)));
-        String destination = String.format("%s/%s", context.workingDir(), config.getValue(FetchTask.DESTINATION));
-        setupDestinationDirectory(destination);
+            String artifactPathOnS3 = fetchConfig.getArtifactsLocationTemplate();
+            context.console().printLine(String.format("Getting artifacts from %s", store.pathString(artifactPathOnS3)));
+            String destination = String.format("%s/%s", context.workingDir(), config.getValue(FetchTask.DESTINATION));
+            setupDestinationDirectory(destination);
 
-        try {
-            store.getPrefix(artifactPathOnS3, destination);
-        } catch (Exception e) {
-            String message = String.format("Failure while downloading artifacts - %s", e.getMessage());
-            logger.error(message, e);
+            try {
+                store.getPrefix(artifactPathOnS3, destination);
+            } catch (Exception e) {
+                String message = String.format("Failure while downloading artifacts - %s", e.getMessage());
+                logger.error(message, e);
+                return ExecutionResult.failure(message, e);
+            }
+
+            ExecutionResult zipArchivesExecutionResult = ExtractAndCleanupZipArchivesIfPresent(destination);
+
+            if (!zipArchivesExecutionResult.isSuccessful()) {
+                return zipArchivesExecutionResult;
+            }
+            else {
+                return ExecutionResult.success("Fetched all artifacts");
+            }
+        }
+        catch (Exception e) {
+            String message = String.format("Error during fetch from s3: %s", e.getMessage());
+            logger.error(message);
             return ExecutionResult.failure(message, e);
         }
+    }
 
-        return ExecutionResult.success("Fetched all artifacts");
+    private ExecutionResult ExtractAndCleanupZipArchivesIfPresent(String destination)
```

Comment:
Sorry - I'm living in the .Net world for the last 10 years, and its naming conventions slip in.

As far as method names, I tend to use longer names if that explains better what the method does (in lieu of comments).  But in case above I do agree the extra info is not useful so I've shortened as suggested.